lib/ethers/multicall.ex

```diff
@@ -0,0 +1,343 @@
+  >
+  ```
+  """
+  @spec aggregate2([
+          Ethers.TxData.t()
+          | {Ethers.TxData.t()}
+          | {Ethers.TxData.t(), aggregate2_options}
+        ]) :: Ethers.TxData.t()
+  def aggregate2(data) when is_list(data) do
+    data
+    |> Enum.map(&aggregate2_encode_data/1)
+    |> Multicall3.aggregate()
+  end
+
+  @doc """
+  Encodes a function call with optional options into a solidity compatible (address,bytes).
+
+  ## Parameters
+  - data: A function call with optional options
+    - to: Overrides the `default_address` (if any) for the respective function call.
+
+  ## Examples
+
+  ```elixir
+  { ContractA.foo(), to: "0x1337..." } |> aggregate2_encode_data()
+  {"0x1337...", false, <<...>>}
+
+  ContractB.bar() |> aggregate2_encode_data()
+  {"0x...", <<...>>}
+  ```
+  """
+  @spec aggregate2_encode_data(
+          Ethers.TxData.t()
+          | {Ethers.TxData.t()}
+          | {Ethers.TxData.t(), aggregate2_options}
+        ) :: {Ethers.Types.t_address(), binary()}
+  def aggregate2_encode_data(data)
+
+  def aggregate2_encode_data({%TxData{data: data}, opts}) do
+    {Keyword.fetch!(opts, :to), hex_decode!(data)}
+  end
+
+  def aggregate2_encode_data({%TxData{data: data, default_address: address}})
+      when not is_nil(address) do
+    {address, hex_decode!(data)}
+  end
+
+  def aggregate2_encode_data(%TxData{data: data, default_address: address})
+      when not is_nil(address) do
+    {address, hex_decode!(data)}
+  end
+
+  @doc """
+  Makes an eth_call to with the given data and overrides using `Ethers.call/2`.
+  The responses are parsed and decoded using `:calls` provided in the overrides.
+  Assumes the call is to Multicall, either from a prepared aggregate2 or aggregate3.
+
+  ## Overrides and Options
+    All from `Ethers.call/2`, and:
+    - `:calls`: List of function calls or selectors. Used for decoding results.
+
+  ## Examples
+
+  ```elixir
+  calls = [ ContractA.foo(), { ContractB.foo(), to: "0x..." } ]
+  calls |> Ethers.Multicall.aggregate3() |> Ethers.Multicall.call(calls: calls)
+  {:ok, [ true: "bar", true: "baz" ]}
+
+  calls = [ ContractA.foo(), { ContractB.foo(), to: "0x..." } ]
+  calls |> Ethers.Multicall.aggregate2() |> Ethers.Multicall.call(calls: calls)
+  {:ok, [ 1337, [ "bar", "baz" ]]}
+  ```
+  """
+  @spec call(Ethers.TxData.t(), Keyword.t()) :: {:ok, any()} | {:error, term()}
+  def call(params, overrides \\ []) do
+    {calls, overrides} = Keyword.pop(overrides, :calls)
+
+    case Ethers.call(params, overrides) do
+      {:ok, result} -> {:ok, decode(result, calls)}
+      {:error, cause} -> {:error, cause}
+    end
+  end
+
+  @doc """
+  Same as `Ethers.Multicall.call/2` but raises on error.
+  """
+  @spec call!(Ethers.TxData.t(), Keyword.t()) :: any() | no_return()
+  def call!(params, overrides \\ []) do
+    case Ethers.Multicall.call(params, overrides) do
+      {:ok, result} -> result
+      {:error, reason} -> raise ExecutionError, reason
+    end
+  end
```

Comment:
I've been thinking about this for a few days and I think it's better if we completely remove `Multicall.call` functions and instead let users know how to use `aggregate3` and `decode` functions in the moduledoc. Something like this:

```elixir
calls = [ContractA.foo(), {ContractB.foo(), to: "0x..."}]

calls
|> Multicall.aggregate3() # Or even aggregate2
|> Ethers.call!()
|> Multicall.decode(calls)
```

What do you think?

Reply:
One other reason for this would be that if we have a custom wrapper for the `call` functions it cannot be used with the upcoming `Ethers.batch/2` in #44